Adding higher order gradients

evcu opened the pull request:
Adding higher-order gradients through small changes.

evcu committed:
Adding higher order gradients

karpathy commented:
Great, will take a look later tonight!

evcu committed:
merging

karpathy commented:
The complexification of the engine is worrying in this case. I haven't fully thought it through yet but I'd wish to support this only if the changes are very minimal, eg changing grad to Value and that's it. Are the rest of things needed? 

Also is there a bug in your derivatives in notebook? Looks like dy/dx^2 is wrong and repeats

evcu committed:
Updating the derrivative markdown

evcu commented:
Updated the markdown cell for the derivative.
 
(1) `zero_grad` is needed since model.paramaters are not the only nodes that accumulate gradients. We have to zero the grad on the entire graph, to ensure correctness. This is even needed without higher gradients. For example, running backward twice with currrent engine gives wrong results.
```
x = Value(3)
y = x*x*x
x.grad=0
y.backward(keep_graph=False, zero_grad=False)
print(x.grad) # prints (27)
x.grad=0
y.backward(keep_graph=False, zero_grad=False)
print(x.grad) # prints (45) incorrect
y.backward(keep_graph=False, zero_grad=True)
print(x.grad) # prints (27) correct
```
I can make zeroing the grad the default behavior and remove the flag. This would reduce complexity I think. Note that we still need to zero grad before backward() call, since if a variable is not connected to the end-node, we find it to have 0 (tf sometimes returns None). 

2) `keep_graph` is there to enable higher-order behavior while keeping the old behavior intact. If we want the higher-order behavior to be the only option, I can remove the flag, so it would be cleaner. I thought it is nice to have the option, so that if the user is not intended to do higher-order gradients they can run the backprop without tracking and that would run faster. Btw `track_backward` might be a better flag name. 

Let me know what you think, I am happy to update it. It's also totally fine if you had something else in mind, then I can drop the PR.

karpathy commented:
"and that would run faster"

I think that ship has sailed a long time ago :)

I want to emphasize cleanliness and interpretability over everything else. Let me stare at this for a bit.

karpathy commented:
Ok preliminary thoughts:

- advise we get rid of flags and make the above behaviors default
- did you perhaps mis `__pow__` changes? Still uses `self.data`

evcu committed:
- Adding tests.
- Addigin pow support.
- Removing flags, making default to track gradient.
- Using pow (x**3) in the notebook example.

evcu committed:
removing unneccessary change in lambda function

evcu commented:
Speed is not a concern?? Let's do sparse layers then :P (already did).

Indeed I missed the pow after sync: update the code and add tests.

karpathy commented:
Ok, hmm I still quite dislike that we have to `x.grad = 0` explicitly.

karpathy commented:
Also using `y = x*x*x` instead of `y = x**3` now gives an error

evcu committed:
adding x.grad=0

evcu commented:
For some reason `grad = 0` line was removed before last call. Added it back. 

One option could be passing the parameters to the backward() call (kind of tf style). Then we can set them 0 under backward() call. Not sure this is better, though it would help to remove the zero grad part from nn.

evcu committed:
Updating notabook notation.

evcu commented:
Hi, Andrej
just tested the code by running the colab's and verified they are working. Training is a bit slower than before since we enabled tracking during backprop by default. 

Quick question: how do you run the tests?

TylerYep commented:
From the README: `python -m pytest`
I tried it just now and it works.

evcu committed:
Updating tests.

evcu commented:
Thanks, Tyler! Tests are passing in this branch, too.

ricardoV94 commented on test/test_engine.py:
```diff
@@ -63,5 +63,50 @@ def test_more_ops():
+    d4y = x.grad
+    assert d4y.data == 0
+
+def test_higher_order():
```
Tests should have different names
